pkg/etcdutil/defrag.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdutil
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// defragData defragments the data directory of each etcd member.
+func defragData(tlsConfig *TLSConfig, etcdConnectionTimeout time.Duration) {
+	client, err := GetTLSClientForEtcd(tlsConfig)
+	if err != nil {
+		logrus.Warnf("failed to create etcd client for defragmentation: %v", err)
+		return
+	}
+
+	for _, ep := range tlsConfig.endpoints {
+		logrus.Infof("Defragmenting etcd member[%s]", ep)
+		ctx, cancel := context.WithTimeout(context.TODO(), etcdConnectionTimeout)
+		status, err := client.Status(ctx, ep)
+		cancel()
+		if err != nil {
+			logrus.Warnf("Failed to get status of etcd member[%s] with error: %v", ep, err)
+			continue
+		}
+		dbSizeBeforeDefrag := status.DbSize
+		ctx, cancel = context.WithTimeout(context.TODO(), etcdConnectionTimeout)
+		_, err = client.Defragment(ctx, ep)
+		cancel()
+		if err != nil {
+			logrus.Warnf("Failed to defragment etcd member[%s] with error: %v", ep, err)
+		} else {
+			logrus.Infof("Finished defragmenting etcd member[%s]", ep)
+		}
+		ctx, cancel = context.WithTimeout(context.TODO(), etcdConnectionTimeout)
+		status, err = client.Status(ctx, ep)
+		cancel()
+		if err != nil {
+			logrus.Warnf("Failed to get status of etcd member[%s] with error: %v", ep, err)
+			continue
```

Comment:
Also, do we need to worry about timeout issues since defrag might be a bit longer running operation?

Reply:
>Is some cleanup required for failed defrag?

Internally defragmentation create `db.tmp` file and take care of deleting it in case of failure. So, no cleanup required cause defragmentation happens at server end.  And anyways, unlike restore call,  defrag is client call on live etcd server. So, it is safe to assume that client need not have any access to server's  etcd data directory.

> Also, do we need to worry about timeout issues since defrag might be a bit longer running operation?
Its already configurable. Do you want me to keep it separate from `etcdConnectionTimeout`?

Reply:
> Its already configurable. Do you want me to keep it separate from etcdConnectionTimeout?

 I was more asking about what happens if the timeout occurs? Would the defrag be completed still? How do we know if the operation times out, if it actually succeeded on the server side or not?